feat: BOPIS feature with OPF CXSPA-10761

Diksha-28 opened the pull request:
for more details please refer : https://wiki.one.int.sap/wiki/display/SAPCCWiki/Path-Finding%3A+support+BOPiS+in+OPF and [CXSPA-10761](https://jira.tools.sap/browse/CXSPA-10761)

Diksha-28 committed:
changes for deliveryMode=pickup

Diksha-28 committed:
test cases

github-actions[bot] committed:
Add license header

Diksha-28 committed:
fixing the cart item count

Commit:
chore(ui): update shipping address and delivery method components as per ui design

Diksha-28 committed:
adding pickup address card

Commit:
changes for billing address and deliverymode for pickup items

Diksha-28 committed:
Merge branch 'develop' into BOPISChange

Commit:
Test cases for BOPIS changes

Diksha-28 committed:
removed comments

Diksha-28 committed:
adding @spartacus/pickup-in-store in dependency

Diksha-28 committed:
Update dependencies.json

Diksha-28 committed:
adding properties to tsconfig.lib.json

to fix __ is missing in compilerOptions.paths.

Commit:
Fix Sonarqube scan error

Diksha-28 committed:
fixing sonarQube error

Diksha-28 committed:
Update user-address.service.ts

Diksha-28 committed:
fixing code formatting error

Commit:
fix dependency order

Matejk00 committed:
Merge branch 'develop' into BOPISChange

raghav-swankier commented on integration-libs/opf/checkout/components/opf-checkout-billing-address-form/opf-checkout-billing-address-form.component.spec.ts:
```diff
@@ -3,17 +3,21 @@
   billingAddress$ = new BehaviorSubject<Address | undefined>(undefined);
   isLoadingAddress$ = new BehaviorSubject<boolean>(false);
   isSameAsDelivery$ = new BehaviorSubject<boolean>(true);
+  pickupNoDefaultAddress$ = new Subject<void>();
```
remove this variable if not used

Diksha-28 replied:
We need to keep pickupNoDefaultAddress$ as the component subscribes to it in ngOnInit. Removing it leads to runtime failures, so it’s required even if not directly used in the test.

Matejk00 commented:
Please resolve the OPF build issues reported in the logs: https://github.com/SAP/spartacus/actions/runs/17727558170/job/50371509411?pr=20674

Matejk00 commented on feature-libs/checkout/base/components/guards/checkout-steps-set.guard.ts:
```diff
@@ -104,6 +106,11 @@ export class CheckoutStepsSetGuard implements OnDestroy {
   }
 
   protected isStepSet(step: CheckoutStep): Observable<GuardResult> {
+    if (!this.hasDeliveryItemsInCart) {
```
Can we move the handling of 'pickup in store' to a guard defined within the OPF checkout sub-library, instead of modifying a guard used by all checkout flows? I believe changing the shared guard could introduce a potential breaking change. Since we only want to handle 'pickup in store' when the OPF flow is active, isolating it in the OPF-specific logic might be a better approach.

ijitendrasap replied:
Now, the setPickupDeliveryMode implemented in the opf-checkout-payment-and-review.component.ts instead of checkout-steps-set.guard.ts

Matejk00 commented on integration-libs/opf/checkout/components/opf-checkout-billing-address-form/opf-checkout-billing-address-form.component.ts:
```diff
@@ -36,8 +40,15 @@ export class OpfCheckoutBillingAddressFormComponent implements OnInit {
+    this.userAddressService.loadAddresses();
+    this.service.setDefaultBillingAddress();
     this.service.getAddresses();
+    this.service.pickupNoDefaultAddress$.subscribe(() => {
```
Could you please verify if this subscription is properly unsubscribed during the flow?

ijitendrasap replied:
Added code to unsubscribe from all subscriptions in ngOnDestroy.

Matejk00 commented on projects/schematics/src/shared/utils/graph-utils_spec.ts:
```diff
@@ -144,6 +144,7 @@ describe('Graph utils', () => {
         SPARTACUS_CHECKOUT,
         SPARTACUS_STOREFINDER,
         SPARTACUS_REQUESTED_DELIVERY_DATE,
+        SPARTACUS_PICKUP_IN_STORE,
```
Is it such change needed?
I am wondering if order of the elements matter in this case.

Diksha-28 replied:
The build was failing, and the logs suggested changing the order. Once we applied this change, the build passed successfully.

Matejk00 commented on integration-libs/opf/base/core/services/opf-endpoints.service.spec.ts:
```diff
@@ -3,12 +3,12 @@
-
 import { BaseSiteService, StringTemplate } from '@spartacus/core';
 import { OpfApiConfig, OpfConfig } from '@spartacus/opf/base/root';
+
```
It seems we can simply revert this file without introducing any changes.

Diksha-28 replied:
done

Matejk00 commented on integration-libs/opf/checkout/components/opf-checkout-billing-address-form/opf-checkout-billing-address-form.service.ts:
```diff
@@ -63,6 +71,13 @@ export class OpfCheckoutBillingAddressFormService {
+  get pickupNoDefaultAddress$(): Observable<void> {
+    return this._pickupNoDefaultAddress$.asObservable();
+  }
+  private handleNoDefaultAddress(): void {
```
We generally avoid using private methods, as they impact customer extensibility. Would you mind changing this function to `protected` instead?

Diksha-28 replied:
sure

raghav-swankier commented on integration-libs/opf/checkout/components/opf-checkout-billing-address-form/opf-checkout-billing-address-form.service.ts:
```diff
@@ -51,6 +56,9 @@ export class OpfCheckoutBillingAddressFormService {
     OpfCheckoutPaymentWrapperService
   );
+  protected userAddressService = inject(UserAddressService);
+  protected _pickupNoDefaultAddress$ = new Subject<void>();
```
Can you use a better naming convention for this? Also, check if you can use BehaviorSubject with an initial value instead of Subject?

ijitendrasap replied:
In payment & checkout page, we cannot initialize it with a default value. Using a BehaviorSubject would require providing an initial value, but at that point we cannot determine whether an address exists in the address book or not. That’s why we are using a plain Subject instead

raghav-swankier commented on integration-libs/opf/checkout/components/opf-checkout-payment-and-review/opf-checkout-payment-and-review.component.html:
```diff
@@ -3,26 +3,6 @@
     value: explicitTermsAndConditions$ | async,
   } as explicitTermsAndConditions"
 >
-  <div class="cx-opf-review-card-wrapper">
```
Please verify and test Hybrid and Shipping mode. I guess these cards should be in place for hybrid and shipping mode.

Diksha-28 replied:
Currently, it works as per the mockup. We’ll perform another round of testing after all review comments are addressed.

raghav-swankier commented on integration-libs/opf/checkout/components/opf-checkout-review-cart-details/opf-checkout-review-cart-details.component.html:
```diff
@@ -13,17 +17,40 @@
+    <div class="cx-items-to-ship-label">
+      {{ 'opfCheckout.itemsToBeShipped' | cxTranslate }}
+    </div>
+    <div class="cx-opf-review-card-wrapper">
```
I see that these cards have been moved here. But verify all the different scenarios and their placement as per the mockup.

Diksha-28 replied:
Currently, it works as per the mockup. I’ll perform another round of testing after all review comments are addressed.

raghav-swankier commented on integration-libs/opf/checkout/components/opf-checkout-review-cart-details/opf-checkout-review-cart-details.component.ts:
```diff
@@ -4,20 +4,22 @@
   standalone: false,
 })
-export class OpfCheckoutReviewCartDetailsComponent {
+export class OpfCheckoutReviewCartDetailsComponent extends CheckoutReviewSubmitComponent {
```
Is it necessary to extend the CheckoutReviewSubmitComponent here?

ijitendrasap replied:
As per the new design, the delivery method and shipping address cards have been moved from (opf-checkout-review-cart-details.component.html) to the review cart section (opf-checkout-review-cart-details.component.html). To support this change, we are extending the CheckoutReviewSubmitComponent, since this class already contains the methods getDeliveryModeCard and getDeliveryAddressCard.

raghav-swankier commented on projects/core/src/user/facade/user-address.service.ts:
```diff
@@ -107,6 +108,15 @@ export class UserAddressService {
+  /**
+   * Returns the default address
+   */
+  getDefaultAddress(): Observable<Address | undefined> {
```
Since this service is not owned by OPF, ask the relevant team to review this part of the code and get their approval as well!

raghav-swankier reviewed (changes requested):
Check the comments

Diksha-28 committed:
addressing review comments.

Diksha-28 committed:
Merge branch 'develop' into BOPISChange

Diksha-28 committed:
Addressing review comments

changed handleNoDefaultAddress method to protected and added optional chaining to address in case address is undefined or null

Commit:
Addressing review comments(deliveryMode)

Commit:
fixed prettier formatting error

Diksha-28 committed:
Merge branch 'develop' into BOPISChange

Commit:
variable rename

Commit:
fixed testcases

Commit:
fixed formatting issue

Diksha-28 committed:
Merge branch 'develop' into BOPISChange

ijitendrasap committed:
Merge branch 'develop' into BOPISChange

Commit:
changes reverted

Diksha-28 committed:
Merge branch 'develop' into BOPISChange

Matejk00 committed:
Merge branch 'develop' into BOPISChange

Commit:
fix: no billing address disables payment options

ijitendrasap committed:
Merge branch 'develop' into BOPISChange

Commit:
build error fix

Commit:
build error fix

Commit:
update dependencies

Commit:
fix: B2B address card, add backtocart , show error when no billing address

Diksha-28 committed:
fixing sonarQube issue

Diksha-28 committed:
static code check fix

Diksha-28 committed:
Merge branch 'develop' into BOPISChange

Commit:
fix: noBillingAddress  error message

Commit:
unit testcases fixed

Commit:
fixed formatting

Commit:
fixed sonarqube and formatting

Commit:
fixed formatting

Matejk00 committed:
Merge branch 'develop' into BOPISChange

Matejk00 commented on feature-libs/checkout/base/assets/translations/en/checkout.json:
```diff
@@ -1,6 +1,7 @@
   "checkout": {
-    "backToCart": "Back to cart"
+    "backToCart": "Back to cart",
+    "back": "Back"
```
I believe we already have this translation key in common.json. Do we need to redefine it specifically for the checkout?

Diksha-28 replied:
Reusing the one in common.json

Matejk00 commented on integration-libs/opf/checkout/components/opf-checkout-billing-address-form/opf-checkout-billing-address-form.component.html:
```diff
@@ -50,9 +50,13 @@ <h3>{{ 'opfCheckout.billingAddress' | cxTranslate }}</h3>
-          cancelBtnLabel="{{ 'common.cancel' | cxTranslate }}"
+          [cancelBtnLabel]="
+            (service.paymentOptionsDisabled$ | async)
+              ? ('checkout.back' | cxTranslate)
```
I think we can also use the translation from `common.back` key.

Matejk00 commented on integration-libs/opf/checkout/components/opf-checkout-billing-address-form/opf-checkout-billing-address-form.component.ts:
```diff
@@ -35,9 +46,25 @@ export class OpfCheckoutBillingAddressFormComponent implements OnInit {
 
   countries$: Observable<Country[]>;
 
+  constructor(
```
I think better can be to use `inject()` instead of constructor injection for more consistency.

ijitendrasap replied:
sure

Matejk00 commented on integration-libs/opf/checkout/components/opf-checkout-billing-address-form/opf-checkout-billing-address-form.component.ts:
```diff
@@ -47,6 +74,17 @@ export class OpfCheckoutBillingAddressFormComponent implements OnInit {
+
+  onBackToAddress(): void {
+    this.subscription.add(
+      this.service.paymentOptionsDisabled$.subscribe((disabled) =>
```
If it's boolean we can re-phrase argument to `isDisabled`

Diksha-28 replied:
sure

Matejk00 commented on integration-libs/opf/checkout/components/opf-checkout-billing-address-form/opf-checkout-billing-address-form.service.ts:
```diff
@@ -74,6 +96,35 @@ export class OpfCheckoutBillingAddressFormService {
+            this.handleNoDefaultAddress();
+          }
+        }),
+        filter((addr): addr is Address => !!addr),
```
Why we need to use `addr is Address` here?

Diksha-28 replied:
We use filter((addr): addr is Address => !!addr) to ensure only valid Address objects are passed to the next operator. This type guard improves type safety, so in the following switchMap, we can confidently call setBillingAddress(defaultAddress) without extra null checks or type assertions. Please let us know if you think there’s a better way or have any other suggestions.

Matejk00 commented on integration-libs/opf/checkout/components/opf-checkout-billing-address-form/opf-checkout-billing-address-form.service.ts:
```diff
@@ -74,6 +96,35 @@ export class OpfCheckoutBillingAddressFormService {
+        take(1),
+        filter((hasDeliveryItems) => !hasDeliveryItems),
+        switchMap(() => this.userAddressService.getDefaultAddress()),
+        tap((defaultAddress) => {
```
Please add argument type.

ijitendrasap replied:
sure

Matejk00 commented on integration-libs/opf/checkout/components/opf-checkout-billing-address-form/opf-checkout-billing-address-form.service.ts:
```diff
@@ -74,6 +96,35 @@ export class OpfCheckoutBillingAddressFormService {
+          }
+        }),
+        filter((addr): addr is Address => !!addr),
+        switchMap((defaultAddress) => {
```
Please add argument type.

ijitendrasap replied:
sure

Matejk00 commented on integration-libs/opf/checkout/components/opf-checkout-payments/opf-checkout-payments.component.html:
```diff
@@ -31,12 +31,21 @@ <h3 *ngIf="isHeadingDisplayed">
                   type="radio"
-                  [attr.aria-checked]="configuration.id === selectedPaymentId"
+                  [attr.aria-checked]="
+                    (paymentDisabled$ | async) &&
```
We can use single async pipe instead of several ones. This reduces subscriptions and keeps the template readable.
Example:
```
<ng-container *ngIf="paymentDisabled$ | async as isPaymentDisabled">
```
then `isPaymentDisabled` can be used everywhere under specific block/element.

ijitendrasap replied:
paymentDisabled$ state individually for specific options, ensuring only the necessary options are disabled. Using a parent-level async pipe would incorrectly disable the entire payment list, which is not the desired behavior

Matejk00 commented on integration-libs/opf/checkout/components/opf-checkout-review-cart-details/opf-checkout-review-cart-details.component.ts:
```diff
@@ -10,19 +10,19 @@ import {
 
   @Input() entries: any[] | null;
-
+  @Input() isAddressCardVisible = false;
```
Can we make this input optional?

Diksha-28 replied:
isAddressCardVisible is already optional because it has a default value (false). If the parent does not provide this input, the component will use the default value and work as expected.

Matejk00 reviewed (changes requested):
I added couple of comments to the PR. There is also error during the build process: https://github.com/SAP/spartacus/actions/runs/18185133357/job/51767950032?pr=20674

@Diksha-28  Can you please verify?

Diksha-28 committed:
addressing review comments

Commit:
addressing review comments

Commit:
fixed formatting

Commit:
fix build error

Diksha-28 committed:
Merge branch 'develop' into BOPISChange

Commit:
fixed error 'cxfeature' can't bind

Matejk00 committed:
chore: add pickup in store lib to be built before OPF

Diksha-28 committed:
Merge branch 'develop' into BOPISChange

Matejk00 commented on feature-libs/estimated-delivery-date/public_api.ts:
```diff
@@ -9,3 +9,4 @@
  */
 export * from './estimated-delivery-date.module';
 export * from './show-estimated-delivery-date/public_api';
+export * from './assets/public_api';
```
Do we need to modify another sublibrary to export its assets in this PR? Is that intentional?

ijitendrasap replied:
No not needed to export

Diksha-28 committed:
Merge branch 'develop' into BOPISChange

github-actions[bot] committed:
Add license header

Matejk00 commented on feature-libs/pickup-in-store/components/container/pickup-items-details/pickup-items-details.module.ts:
```diff
@@ -28,6 +29,7 @@ import { PickUpItemsDetailsComponent } from './pickup-items-details.component';
     StoreModule,
     CardModule,
     MediaModule,
+    FeaturesConfigModule,
```
Do we need this import? Because this PR do not introduce any changes in the TS file or the template.

ijitendrasap replied:
The module declares and exports [FeatureDirective], which enables the *cxFeature structural directive in templates. Importing this module resolves the error NG0303: Can't bind to 'cxFeature' since it isn't a known property of 'div' in _PickUpItemsDetailsComponent.

Matejk00 commented on integration-libs/opf/b2b-checkout/components/opf-b2b-checkout-payment-and-review/opf-b2b-checkout-payment-and-review.component.html:
```diff
@@ -110,5 +110,6 @@ <h3>{{ 'opfCheckout.termsAndConditions' | cxTranslate }}</h3>
   <cx-opf-checkout-review-cart-details
     [cart]="cart$ | async"
     [entries]="entries$ | async"
+    [isAddressCardVisible]="false"
```
Can we remove this line as `isAddressCardVisible` is `false` by default?

ijitendrasap replied:
yes

Matejk00 commented on integration-libs/opf/b2b-checkout/components/opf-b2b-checkout-review/opf-b2b-checkout-review.component.html:
```diff
@@ -110,6 +110,7 @@ <h3>{{ 'opfCheckout.termsAndConditions' | cxTranslate }}</h3>
   <cx-opf-checkout-review-cart-details
     [cart]="cart$ | async"
     [entries]="entries$ | async"
+    [isAddressCardVisible]="false"
```
Can we remove this line as `isAddressCardVisible` is `false` by default?

ijitendrasap replied:
yes

Matejk00 commented on integration-libs/opf/checkout/components/opf-checkout-billing-address-form/opf-checkout-billing-address-form.service.spec.ts:
```diff
@@ -251,4 +269,65 @@ describe('OpfCheckoutBillingAddressFormService', () => {
+    });
+    (service as any).handleNoDefaultAddress();
+  });
+  it('should handle error when setting default billing address fails', fakeAsync(() => {
```
Nitpick: We can add empty line between these test scenarios.

ijitendrasap replied:
sure

Matejk00 commented on integration-libs/opf/checkout/components/opf-checkout-billing-address-form/opf-checkout-billing-address-form.service.spec.ts:
```diff
@@ -251,4 +269,65 @@ describe('OpfCheckoutBillingAddressFormService', () => {
+
+    expect(handleNoDefaultAddressSpy).toHaveBeenCalled();
+  }));
+  it('should handle errors when loading the default address', fakeAsync(() => {
```
Nitpick: We can add empty line between these test scenarios.

ijitendrasap replied:
sure

Matejk00 commented on projects/core/src/util/glob-utils.ts:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2010-2019 Google LLC. http://angular.io/license
+ * Copyright (C) 2010-2019 Google LLC. http://angular.io/license
```
I think is unintended change

ijitendrasap replied:
these changes automatically done by github-actions[bot](Add license header)

Matejk00 commented on projects/schematics/src/shared/utils/html-utils.ts:
```diff
@@ -2,8 +2,8 @@ import { SchematicsException, Tree } from '@angular-devkit/schematics';
 /*
- * Copyright Google LLC All Rights Reserved.
  * SPDX-FileCopyrightText: 2025 SAP Spartacus team <spartacus-team@sap.com>
+ * Copyright Google LLC All Rights Reserved.
```
Please revert this, we do not need to modify this in such PR.

ijitendrasap replied:
these changes automatically done by github-actions[bot](Add license header)

Matejk00 commented on projects/schematics/src/shared/utils/load-esm-module.ts:
```diff
@@ -6,8 +6,8 @@
 /*
- * Copyright Google LLC All Rights Reserved.
  * SPDX-FileCopyrightText: 2025 SAP Spartacus team <spartacus-team@sap.com>
+ * Copyright Google LLC All Rights Reserved.
```
Please revert this, we do not need to modify this in such PR.

ijitendrasap replied:
these changes automatically done by github-actions[bot](Add license header)

Matejk00 commented on projects/schematics/src/shared/utils/project-tsconfig-paths.ts:
```diff
@@ -1,6 +1,6 @@
 /*
- * Copyright Google LLC All Rights Reserved.
  * SPDX-FileCopyrightText: 2025 SAP Spartacus team <spartacus-team@sap.com>
+ * Copyright Google LLC All Rights Reserved.
```
Please revert this, we do not need to modify this in such PR.

ijitendrasap replied:
these changes automatically done by github-actions[bot](Add license header)

Commit:
addressing review comments

github-actions[bot] committed:
Add license header

Diksha-28 committed:
Merge branch 'develop' into BOPISChange

raghav-swankier commented on integration-libs/opf/checkout/components/opf-checkout-payment-and-review/opf-checkout-payment-and-review.component.ts:
```diff
@@ -100,7 +104,20 @@ export class OpfCheckoutPaymentAndReviewComponent
+      .subscribe((hasDeliveryItems) => {
+        if (!hasDeliveryItems) {
+          this.checkoutDeliveryAddressFacade.clearCheckoutDeliveryAddress();
+          this.checkoutDeliveryModesFacade.setDeliveryMode('pickup');
```
Better to use enum or constant here instead of string directly

ijitendrasap replied:
As No enums were found for delivery modes,Similar usage without enums can be seen in checkout-steps-set.guard.ts.

raghav-swankier commented on integration-libs/opf/checkout/components/opf-checkout-payments/opf-checkout-payments.component.spec.ts:
```diff
@@ -115,6 +117,10 @@ describe('OpfCheckoutPaymentsComponent', () => {
       of(mockOpfMetadata)
     );
+    mockBillingAddressFormService = {
+      paymentOptionsDisabled$: of(false), // Mock the observable
```
you can remove the comment

ijitendrasap replied:
sure

raghav-swankier commented on integration-libs/opf/checkout/styles/components/_opf-checkout-payment-and-review.scss:
```diff
@@ -122,14 +122,15 @@
 
   %edit-button-styles {
-    position: absolute;
+    position: center;
```
is this change in css necessary?

ijitendrasap replied:
Yes, the position was changed to center alignment to correct button alignment and resolve overflow issues.

raghav-swankier reviewed (dismissed):
Check if the minor things could be addressed in this PR itself.

Commit:
addressing review comments

ijitendrasap committed:
Merge branch 'develop' into BOPISChange

ijitendrasap committed:
Merge branch 'develop' into BOPISChange

ijitendrasap committed:
Merge branch 'develop' into BOPISChange

Commit:
fix dependencies error

Commit:
fix dependency order

Matejk00 committed:
Merge branch 'develop' into BOPISChange
